### PR TITLE
Add the `viz` add-ons to `kedro new`

### LINKED ISSUE #3074: Add `kedro-viz` to add-ons options
## Description
`Kedro Viz` should be added to the add-ons options, allowing users to get setup for viz (see https://github.com/kedro-org/kedro/issues/2506#issuecomment-1717432995). This option should make use of the new `spaceflights-pandas-viz` starter: https://github.com/kedro-org/kedro/issues/2985

When a user selects `viz` they should get the config/requirements needed to run Kedro viz on their project. This shouldn't include the example pipelines. This will happen in a follow up ticket. 

## Possible Implementation
This solution is likely going to be the same as for the other `add-ons` work, leveraging the post-project hooks. Since we'd use the spaceflights-pandas-viz starter and not the regular template, we probably need to copy some of the logic from the post-project hook in the regular template to the starter. 

## Possible Alternatives
N/A

### PR DESCRIPTION
**Pull request by noklam**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Fix #3074 and #3075

## Development notes

@SajidAlamQB Notes:

- Refactored `utils.py` and unstripped `pyproject.toml` and `requirements.txt`. We now handle removing headers from `pyproject.toml` using the `toml` library.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by noklam** [comment]
I think this should be merged first - https://github.com/kedro-org/kedro/pull/3230. The current test suite is still asserting text, we should read the parse the file.

**Comment by merelcht** [comment]
One thought I had while looking at the code is whether we can leverage the existing `delete_pipeline` functionality to remove the pipeline and parameter files: https://github.com/kedro-org/kedro/blob/main/kedro/framework/cli/pipeline.py#L143

File: kedro/templates/project/hooks/utils.py [hunk 1]
```diff
@@ -1,133 +1,139 @@
...
+            file_path.unlink()
+
+    # Empty the contents of conf/base/catalog.yml
+    catalog_yml_path = current_dir / "conf/base/catalog.yml"
+    if catalog_yml_path.exists():
+        catalog_yml_path.write_text('')
+    # Remove parameter/reporting files from conf/base
+    conf_base_path = current_dir / "conf/base/"
+    if "Kedro Viz" in selected_add_ons_list:  # Remove reporting.yml if Kedro Viz is selected
+        param_to_remove = ["parameters_data_processing.yml", "parameters_data_science.yml", "parameters_reporting.yml", "reporting.yml"]
```
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/templates/project/hooks/utils.py:91]: It should be named `parameters_reporting.yml`, I made a PR for that https://github.com/kedro-org/kedro-starters/pull/178

File: kedro/templates/project/hooks/utils.py [hunk 2]
```diff
@@ -1,133 +1,139 @@
...
+    # Keep lines that are not in content_to_remove
+    lines = [line for line in lines if line.strip() not in content_to_remove_lines]
+
+    with open(file_path, 'w') as file:
+        file.writelines(lines)
+
+
+def remove_nested_section(data, nested_key):
+    keys = nested_key.split('.')
+    d = data
```
[the remark this answers is not in the document]
Comment (SajidAlamQB) [reply kedro/templates/project/hooks/utils.py:33]: Its essentially removing the nested sections of the `pyproject.toml` requirements and removing the parent header once its empty.

File: kedro/framework/cli/starters.py [hunk 3]
```diff
@@ -511,11 +523,17 @@ def _make_cookiecutter_args(
...
+        cookiecutter_args["directory"] = "spaceflights-pyspark-viz"
+    elif add_ons and "Pyspark" in add_ons:
         cookiecutter_args["directory"] = "spaceflights-pyspark"
-        pyspark_path = "git+https://github.com/kedro-org/kedro-starters.git"
-        return pyspark_path
-    return template_path
+    elif add_ons and "Kedro Viz" in add_ons:
+        cookiecutter_args["directory"] = "spaceflights-pandas-viz"
+    else:
+        starter_path = template_path
```
[the remark this answers is not in the document]
Comment (SajidAlamQB) [reply kedro/framework/cli/starters.py:535]: I tried adding this but tests started failing. It should have worked but there might be some underlying issues here.

**Comment by merelcht** [comment]
Should https://github.com/kedro-org/kedro-starters/blob/main/spaceflights-pyspark/hooks/post_gen_project.py be updated to also mention "Kedro Viz" not in addons? The opposite of https://github.com/kedro-org/kedro-starters/blob/main/spaceflights-pandas-viz/hooks/post_gen_project.py

File: kedro/templates/project/hooks/utils.py [hunk 4]
```diff
@@ -1,133 +1,139 @@
...
-            shutil.rmtree(str(tests_path))
+    # Remove the pipelines subdirectories
+    if "Kedro Viz" in selected_add_ons_list: # Remove reporting if Kedro Viz is selected
+        pipelines_to_remove = ["data_science", "data_processing", "reporting"]
     else:
-        with open(requirements_file_path, 'a') as file:
-            file.write(test_requirements)
-        with open(pyproject_file_path, 'a') as file:
-            file.write(test_pyproject_requirements)
+        pipelines_to_remove = ["data_science", "data_processing"]
```
[the remark this answers is not in the document]
Comment (SajidAlamQB) [reply kedro/templates/project/hooks/utils.py:101]: The core template has a `pipelines` directory that isn't removed through any of the add-ons options. We'll have to remove that as well to keep starters and core aligned. I think we can just remove the pipelines directory manually for core though?

**Comment by SajidAlamQB** [comment]
> Should https://github.com/kedro-org/kedro-starters/blob/main/spaceflights-pyspark/hooks/post_gen_project.py be updated to also mention "Kedro Viz" not in addons? The opposite of https://github.com/kedro-org/kedro-starters/blob/main/spaceflights-pandas-viz/hooks/post_gen_project.py

Nice spot! I don't think it really matters as the templates are decided in `starters.py` so where the add-ons work is concerned its not a problem. This was more of guard for users using the `--starter` flags as the default add-on in starters is `none` so as long as we have a check for any other add-on it won't run. But for consistency I think it would be nice to make them all aligned - it’s cleaner that way!

File: kedro/framework/cli/starters.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SajidAlamQB) [reply kedro/framework/cli/starters.py]: Not a problem, managed to fix it. 👍

File: tests/framework/cli/test_starters.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 import shutil
 from pathlib import Path
 
+import anyconfig
```
Comment (merelcht) [inline tests/framework/cli/test_starters.py:8]: @ankatiyar just got rid of `anyconfig` in #2692 , can we use `toml` or `omegaconf` here instead as well?

File: kedro/templates/project/hooks/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/templates/project/hooks/utils.py]: That's a really good point. Let's park that for now and I'll have a discussion with the Alloy team first on expectations.

**Review by merelcht — COMMENTED** [review]
I've tried this out and the behaviour looks good! I found a coupling between the existence of the `data` folder and the `SQLiteSessionStore` config, but let's decide on that via Slack. 

The only thing that's missing is just the docstring for the helper methods, but happy to approve when that's done 👍

File: kedro/templates/project/hooks/utils.py [hunk 6]
```diff
@@ -1,133 +1,137 @@
...
+
+
+def remove_dir(path):
+    if path.exists():
+        shutil.rmtree(str(path))
+
+
+def remove_file(path):
+    if path.exists():
+        path.unlink()
```
Comment (noklam) [inline kedro/templates/project/hooks/utils.py:74]: minor question should these all be private methods? I think only the `selected_template_add_ons` should be public, but this is minor and we can address this separately in  https://github.com/kedro-org/kedro/issues/3264

**Comment by noklam** [comment]
I have fix the tests with `toml` and add some docstring for the new utils question. I leave a question whether these should be all public methods and maybe worth refactor later but not necessary to fix this in this PR.

Documented my observation in https://github.com/kedro-org/kedro/issues/3264

File: kedro/templates/project/hooks/utils.py (hunk 6 above)
Comment (SajidAlamQB) [reply kedro/templates/project/hooks/utils.py:74]: Yes they should be, only `setup_template_add_ons` and `sort_requirements` should be public.

Comment (SajidAlamQB) [reply kedro/templates/project/hooks/utils.py]: Made them private.

File: kedro/templates/project/hooks/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro/templates/project/hooks/utils.py]: Update: discussed this with @yetudada and @amandakys and we decided that the empty `pipelines` directory should stay.

File: kedro/templates/project/hooks/utils.py [hunk 7]
```diff
@@ -1,140 +1,189 @@
...
+    raw_data_path = current_dir / "data/01_raw/"
+    for file_path in raw_data_path.glob("*.*"):
+        if file_path.suffix in [".csv", ".xlsx"]:
+            file_path.unlink()
+
+    # Empty the contents of conf/base/catalog.yml
+    catalog_yml_path = current_dir / "conf/base/catalog.yml"
+    if catalog_yml_path.exists():
+        catalog_yml_path.write_text('')
+    # Remove parameter/reporting files from conf/base
```
Comment (merelcht) [inline kedro/templates/project/hooks/utils.py:125]:
```suggestion
    # Remove parameter files from conf/base
```

**Review by merelcht — APPROVED** [review]
LGTM! Great work @SajidAlamQB and @noklam ⭐ ⭐ ⭐

File: tests/framework/cli/test_starters.py [hunk 8]
```diff
@@ -89,12 +91,13 @@ def _convert_addon_names_to_numbers(selected_addons: str):
 
 def _get_expected_files(add_ons: str):
     add_ons_template_files = {
-        "1": 0,
-        "2": 3,
-        "3": 1,
-        "4": 2,
-        "5": 8,
-        "6": 3,
+        "1": 0,  # Linting does not add any files
```
Comment (AhdraMeraliQB) [inline tests/framework/cli/test_starters.py:94]: Smart choice adding the comments!

**Review by AhdraMeraliQB — APPROVED** [review]
All looks good to me! Good work refactoring the tests 🌟

**Pull request merged**